### PR TITLE
[MDB IGNORE] Fix ticket machines behaving unexpectedly and causing runtimes

### PR DESCRIPTION
**Pull request by strandsofivy**
## About The Pull Request

The ticket machine used to desynchronize very easily due to out of bounds list accesses when trying to access the current ticket. The list that keeps track of tickets has been repurposed more into a queue, pushing all elements to the front with .Cut(1,2), compared to before where it would leave a space of nothingness where tickets used to be. 

Also, there was a ticket machine in a ruin in icebox which had the default id, causing an id collision.

## Why It's Good For The Game

bgug fix

## Changelog

:cl:
fix: HoPs all around the Spinward Sector scream in joy as their beloved ticket machine finally rises back from the dead.
qol: Add more examines to ticket machines and tickets, giving info about people's spot in queue and so on.
/:cl:

### CODE REVIEW

File: code/modules/paperwork/ticketmachine.dm [hunk 1]
```diff
@@ -219,10 +234,24 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
...
 	resistance_flags = FLAMMABLE
 	max_integrity = 50
+	var/number
 	var/saved_maptext = null
 	var/owner_ref // A ref to our owner. Doesn't need to be weak because mobs have unique refs
 	var/obj/machinery/ticket_machine/source
 
+/obj/item/ticket_machine_ticket/New(loc, number)
+	. = ..()
+	src.number = number
```
Comment (ShizCalev) [inline code/modules/paperwork/ticketmachine.dm:244]: name the arg something else, src.number makes me very sad.

Comment (ShizCalev) [inline code/modules/paperwork/ticketmachine.dm:244]: name the arg something else, src.number makes me very sad.

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by distributivgesetz** [comment]
what

File: code/modules/paperwork/ticketmachine.dm [hunk 2]
```diff
@@ -126,6 +144,8 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
 		return
 	cooldown = TRUE
 	machine.increment()
+	if(machine.current_ticket == null)
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:147]:
```suggestion
	if(isnull(machine.current_ticket))
```

File: code/modules/paperwork/ticketmachine.dm [hunk 3]
```diff
@@ -208,6 +225,7 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
 		user.adjust_fire_stacks(1)
 		user.ignite_mob()
 		return
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:227]: Should emagged ones also call ``update_appearance()``? Maybe we can just remove this return if so.

File: code/modules/paperwork/ticketmachine.dm [hunk 4]
```diff
@@ -219,10 +237,26 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
...
+	. = ..()
+	number = num
+	if(number != null) // dont make a zero ticket or ill cry
+		name += " #[number]"
+		saved_maptext = MAPTEXT(number)
+		maptext = saved_maptext
+
+/obj/item/ticket_machine_ticket/examine(mob/user)
+	. = ..()
+	if(number != null)
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:255]:
```suggestion
	if(!isnull(number))
```

File: code/modules/paperwork/ticketmachine.dm [hunk 5]
```diff
@@ -233,6 +267,7 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
 
 	return ..()
 
+// ???????????????
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:270]: maybe you can move this piece of code to paper?

File: code/modules/paperwork/ticketmachine.dm [hunk 6]
```diff
@@ -98,7 +114,8 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
 /obj/item/assembly/control/ticket_machine
 	name = "ticket machine controller"
 	desc = "A remote controller for the HoP's ticket machine."
-	var/datum/weakref/linked //To whom are we linked?
+	///To whom are we linked?
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:117]: Linked to what? please clarify here

File: code/modules/paperwork/ticketmachine.dm [hunk 7]
```diff
@@ -98,7 +114,8 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
...
+	var/datum/weakref/linked
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:118]: weakref vars should preferably end in ``_ref`` so it's easier to know they're a weakref, please do so if you have the time.

File: code/modules/paperwork/ticketmachine.dm [hunk 8]
```diff
@@ -190,11 +210,8 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
 		return
 	playsound(src, 'sound/machines/terminal_insert_disc.ogg', 100, FALSE)
 	ticket_number++
-	to_chat(user, span_notice("You take a ticket from [src], looks like you're ticket number #[ticket_number]..."))
-	var/obj/item/ticket_machine_ticket/theirticket = new /obj/item/ticket_machine_ticket(get_turf(src))
-	theirticket.name = "Ticket #[ticket_number]"
-	theirticket.maptext = MAPTEXT(ticket_number)
-	theirticket.saved_maptext = MAPTEXT(ticket_number)
+	to_chat(user, span_notice("You take a ticket from [src], looks like you're number [ticket_number] in queue..."))
+	var/obj/item/ticket_machine_ticket/theirticket = new /obj/item/ticket_machine_ticket(get_turf(src), ticket_number)
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:214]:
```suggestion
	var/obj/item/ticket_machine_ticket/theirticket = new (get_turf(src), ticket_number)
```

File: code/modules/paperwork/ticketmachine.dm [hunk 9]
```diff
@@ -219,10 +237,26 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
...
 	var/owner_ref // A ref to our owner. Doesn't need to be weak because mobs have unique refs
 	var/obj/machinery/ticket_machine/source
 
+/obj/item/ticket_machine_ticket/New(loc, num)
+	. = ..()
+	number = num
+	if(number != null) // dont make a zero ticket or ill cry
+		name += " #[number]"
+		saved_maptext = MAPTEXT(number)
+		maptext = saved_maptext
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:251]: This should be in ``Initialize()``

File: code/modules/paperwork/ticketmachine.dm [hunk 10]
```diff
@@ -219,10 +237,26 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
...
 	max_integrity = 50
+	var/number
 	var/saved_maptext = null
 	var/owner_ref // A ref to our owner. Doesn't need to be weak because mobs have unique refs
 	var/obj/machinery/ticket_machine/source
 
+/obj/item/ticket_machine_ticket/New(loc, num)
+	. = ..()
+	number = num
+	if(number != null) // dont make a zero ticket or ill cry
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:248]:
```suggestion
	if(!isnull(number)) // dont make a zero ticket or ill cry
```

File: code/modules/paperwork/ticketmachine.dm [hunk 11]
```diff
@@ -219,10 +233,26 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/ticket_machine, 32)
 	w_class = WEIGHT_CLASS_TINY
 	resistance_flags = FLAMMABLE
 	max_integrity = 50
+	var/number
 	var/saved_maptext = null
 	var/owner_ref // A ref to our owner. Doesn't need to be weak because mobs have unique refs
 	var/obj/machinery/ticket_machine/source
 
+/obj/item/ticket_machine_ticket/Initialize(loc, num)
```
Comment (JohnFulpWillard) [inline code/modules/paperwork/ticketmachine.dm:241]:
```suggestion
/obj/item/ticket_machine_ticket/Initialize(mapload, num)
```

**Comment by JohnFulpWillard** [comment]
Thank you for all your work

**Comment by distributivgesetz** [comment]
Thanks for the fast reviews, I swear I'll learn this language someday lmao

**Pull request merged**